src/app.js

```diff
@@ -25,107 +25,92 @@ export function app(props) {
+    copy(state, module.state)
+
+    setupActions(actions, state, module.actions)
+
+    if (module.init) {
+      callbackStack.push(function() {
+        module.init(state, actions)
+      })
+    }
+
+    if (module.modules) {
+      Object.keys(module.modules).map(function(key) {
+        setup(module.modules[key], (state[key] = {}), (actions[key] = {}))
+      })
+    }
   }
 
-  function createActions(actions, withActions, lastPath) {
+  function setupActions(actions, withState, withActions) {
     Object.keys(withActions || {}).map(function(name) {
-      return typeof withActions[name] === "function"
-        ? (actions[name] = function(data) {
-            return typeof (data = withActions[name](
-              getPath(lastPath, appState),
-              getPath(lastPath, appActions),
-              data
-            )) === "function"
-              ? data(update)
-              : update(data)
-          })
-        : createActions(
-            actions[name] || (actions[name] = {}),
-            withActions[name],
-            lastPath.concat(name)
-          )
+      if (typeof withActions[name] === "function") {
+        actions[name] = function(data) {
+          return typeof (data = withActions[name](withState, actions, data)) ===
+          "function"
+            ? data(update)
+            : update(data)
+        }
+      } else {
+        setupActions(
+          (actions[name] = {}),
+          withState[name] || (withState[name] = {}),
+          withActions[name]
+        )
+      }
     })
 
-    function update(withState) {
-      if (typeof withState === "function") {
-        return update(withState(getPath(lastPath, appState)))
-      }
-      if (
-        withState &&
-        (withState = setPath(
-          lastPath,
-          merge(getPath(lastPath, appState), withState),
-          appState
-        ))
-      ) {
-        requestRender((appState = withState))
-      }
-      return appState
+    function update(data) {
+      return (
+        typeof data === "function"
+          ? update(data(withState))
+          : data && requestRender(copy(withState, data)),
+        withState
+      )
     }
   }
 
-  function set(prop, value, source) {
-    var target = merge(source)
-    target[prop] = value
+  function copy(target, source) {
+    for (var i in source) {
+      target[i] = source[i]
+    }
     return target
   }
 
-  function getPath(paths, source) {
-    return paths.length === 0
-      ? source
-      : source && getPath(paths.slice(1), source[paths[0]])
-  }
-
-  function setPath(paths, value, source) {
-    var name = paths[0]
-    return paths.length === 0
-      ? value
-      : set(
-          name,
-          paths.length > 1
-            ? setPath(
-                paths.slice(1),
-                value,
-                source && name in source ? source[name] : {}
-              )
-            : value,
-          source
-        )
+  function merge(target, source, result) {
```

Comment:
It looks like this is only ever called with the `target` and `source`, and `result` is always set to `{}`. Will this work just as well?

```js
function merge(target, source) {
  return copy(copy({}, target), source)
}
```

All the tests pass, but for some reason the gzip output is a byte larger 😕

Reply:
Good idea, this should work.